lua/ulx/modules/urs_server.lua

```diff
@@ -1,193 +1,268 @@
+
+    if echoSpawns then
+        logSpawn(ply:Nick() .."<".. ply:SteamID() .."> spawned/used ".. restrictionType .." ".. what .." -=RESTRICTED=-")
+    end
+    ULib.tsayError(ply, "\"".. what .."\" is a restricted ".. restrictionType .." from your rank.")
+end
+local PrintRestricted = URS.PrintRestricted
+
+-- Return cached value for this check
+function URS.CachedCheck(ply, restrictionType, what)
+    ply.URS_CacheCheck = ply.URS_CacheCheck or {}
+
+    local existing = ply.URS_CacheCheck
+    local restrictionTypes = rawget( existing, restrictionType ) 
+
+    if not restrictionTypes then return end
+
+    -- true, false, or nil
+    return rawget( restrictionTypes, what )
 end
+local cachedCheck = URS.CachedCheck
+
+-- Caches the given check and returns the given result
+function URS.CacheCheck(ply, restrictionType, what, result)
+    ply.URS_CacheCheck = ply.URS_CacheCheck or {}
+
+    local existing = ply.URS_CacheCheck
+    local restrictionTypes = rawget( existing, restrictionType ) 
+
+    if not restrictionTypes then
+        rawset( existing, restrictionType, {} )
+    end
 
-function URS.PrintRestricted(ply, type, what) 
-	if type == "pickup" then return end -- Constant spam
-	if URS.cfg.echoSpawns:GetBool() then 
-		ulx.logSpawn(ply:Nick() .."<".. ply:SteamID() .."> spawned/used ".. type .." ".. what .." -=RESTRICTED=-")
-	end 
-	ULib.tsayError(ply, "\"".. what .."\" is a restricted ".. type .." from your rank.")
-end 
-
-function URS.Check(ply, type, what)
-	what = string.lower(what) 
-	local group = ply:GetUserGroup() 
-	local restriction = false
-
-	if URS.restrictions[type] and URS.restrictions[type][what] then 
-		restriction = URS.restrictions[type][what] 
-	end 
-
-	if restriction then 
-		if table.HasValue(restriction, "*") then 
-			if !(table.HasValue(restriction, group) or table.HasValue(restriction, ply:SteamID())) then 
-				URS.PrintRestricted(ply, type, what)
-				return false
-			end 
-		elseif table.HasValue(restriction, group) or table.HasValue(restriction, ply:SteamID()) then 
-			URS.PrintRestricted(ply, type, what) 
-			return false 
-		end 
-	end 
-	
-	if URS.restrictions["all"] and URS.restrictions["all"][type] and table.HasValue(URS.restrictions["all"][type], group) then 
-		ULib.tsayError(ply, "Your rank is restricted from all ".. type .."s") 
-		return false 
-	elseif table.HasValue(URS.types.limits, type) and URS.limits[type] and (URS.limits[type][ply:SteamID()] or URS.limits[type][group]) then 
-		if URS.limits[type][ply:SteamID()] then 
-			if ply:GetCount(type.."s") >= URS.limits[type][ply:SteamID()] then 
-				ply:LimitHit( type .."s" )
-				return false 
-			end 
-		elseif URS.limits[type][group] then 
-			if ply:GetCount(type.."s") >= URS.limits[type][group] then 
-				ply:LimitHit( type .."s" )
-				return false 
-			end 
-		end 
-		if URS.cfg.overwriteSbox:GetBool() then 
-			return true -- Overwrite sbox limit (ours is greater) 
-		end 
-	end 
+    rawset( restrictionTypes, what, result )
```

Comment:
[urs] addons/urs/lua/ulx/modules/urs_server.lua:92: bad argument #1 to 'rawset' (table expected, got nil)
  1. rawset - [C]:-1
   2. Check - addons/urs/lua/ulx/modules/urs_server.lua:92
    3. fn - addons/urs/lua/ulx/modules/urs_server.lua:239
     4. Call - lua/ulib/shared/hook.lua:109
      5. unknown - gamemodes/sandbox/gamemode/commands.lua:845
       6. unknown - lua/includes/modules/concommand.lua:54

       when spawning in a restricted weapon, only happens once

Reply:
the print only prints once ""weapon_pistol" is a restricted swep from your rank." and doesn't print after trying to spawn anymore

Reply:
Fixed by returning early